File: leo2-central/src/main/java/org/deku/leo2/central/data/repositories/jooq/NodeJooqRepository.java
package org.deku.leo2.central.data.repositories.jooq;

import org.deku.leo2.central.data.entities.jooq.Tables;
import org.deku.leo2.central.data.entities.jooq.tables.MstNode;
import org.deku.leo2.central.data.entities.jooq.tables.records.MstNodeRecord;
import org.jooq.DSLContext;

import javax.inject.Inject;
import javax.inject.Named;

/**
 * Created by masc on 02.07.15.
 */
@Named
public class NodeJooqRepository {
    @Inject
    DSLContext mDSLContext;

    /**
     * Find node record by key
     * @param key Key
     * @return Node record
     */
    public MstNodeRecord findByKey(String key) {
        return mDSLContext.fetchOne(MstNode.MST_NODE, Tables.MST_NODE.KEY.eq(key));
    }

    /**
     * Save record by key
     * @param key
     * @param hardwareAddress
     * @param systemInfo
     */
    public MstNodeRecord saveByKey(String key, String hardwareAddress, String systemInfo) {
        MstNodeRecord r = this.findByKey(key);
        if (r == null)
            r = mDSLContext.newRecord(Tables.MST_NODE);

        r.setKey(key);
        r.setHostname(hardwareAddress);
        r.setSysInfo(systemInfo);
        r.store();

        return r;
    }
}

File: leoz-central/src/main/java/org/deku/leoz/central/data/repositories/NodeJooqRepository.kt
package org.deku.leoz.central.data.repositories

import org.deku.leoz.central.data.entities.jooq.Tables
import org.deku.leoz.central.data.entities.jooq.tables.MstNode
import org.deku.leoz.central.data.entities.jooq.tables.records.MstNodeRecord
import org.jooq.DSLContext

import javax.inject.Inject
import javax.inject.Named

/**
 * Created by masc on 02.07.15.
 */
@Named
class NodeJooqRepository {
    @Inject
    private lateinit var dslContext: DSLContext

    /**
     * Create new record
     */
    fun createNew(): MstNodeRecord {
        return dslContext.newRecord(Tables.MST_NODE)
    }

    /**
     * Find node record by key
     * @param key Key
     * @return Node record
     */
    fun findByKey(key: String): MstNodeRecord? {
        return dslContext.fetchOne(MstNode.MST_NODE, Tables.MST_NODE.KEY.eq(key))
    }

    /**
     * Find node record by key starting with
     * @param partialKey Partial key
     */
    fun findByKeyStartingWith(partialKey: String): MstNodeRecord? {
        return dslContext.fetchOne(MstNode.MST_NODE, Tables.MST_NODE.KEY.startsWith(partialKey))
    }

    /**
     * Check if there's a record for a specific key
     * @param key
     * @return
     */
    fun hasAuthorizedKey(key: String): Boolean {
        val r = this.findByKey(key)
        return r != null && r.authorized !== 0
    }

    /**
     * Find node record by key or create new one if there's none
     * @param key
     */
    fun findByKeyOrCreateNew(key: String): MstNodeRecord {
        var r = this.findByKey(key)
        return r ?: dslContext.newRecord(Tables.MST_NODE)
    }
}
